src/gridcoin/voting/registry.cpp

```diff
@@ -311,6 +317,174 @@ int64_t PollReference::Expiration() const
     return m_timestamp + (m_duration_days * 86400);
 }
 
+CBlockIndex* PollReference::GetStartingBlockIndexPtr() const
+{
+    uint256 block_hash;
+    CTransaction tx;
+
+    GetTransaction(*m_ptxid, tx, block_hash);
+
+    if (!block_hash.IsNull()) {
+        CBlockIndex* pindex = mapBlockIndex[block_hash];
+
+        return pindex;
+    }
+
+    return nullptr;
+}
+
+CBlockIndex* PollReference::GetEndingBlockIndexPtr() const
+{
+    // Has poll ended?
+    if (Expiration() < GetAdjustedTime()) {
+        GRC::BlockFinder blockfinder;
+
+        // Find the last block that contains valid votes for the poll.
+        CBlockIndex* pindex = blockfinder.FindByMinTime(Expiration());
+
+        return pindex;
+    }
+
+    return nullptr;
+}
+
+std::optional<int> PollReference::GetStartingHeight() const
+{
+    const CBlockIndex* pindex = GetStartingBlockIndexPtr();
+
+    if (pindex != nullptr) {
+        return pindex->nHeight;
+    }
+
+    return std::nullopt;
+}
+
+std::optional<int> PollReference::GetEndingHeight() const
+{
+    CBlockIndex* pindex = GetEndingBlockIndexPtr();
+
+    if (pindex != nullptr) {
+        return pindex->nHeight;
+    }
+
+    return std::nullopt;
+}
+
+std::optional<CAmount> PollReference::GetActiveVoteWeight() const
+{
+    // Instrument this so we can log real time performance.
+    g_timer.InitTimer(__func__, LogInstance().WillLogCategory(BCLog::LogFlags::VOTE));
+
+    // Get the start and end of the poll.
+    CBlockIndex* pindex_start = GetStartingBlockIndexPtr();
+    const CBlockIndex* pindex_end = GetEndingBlockIndexPtr();
+
+    // If both start and end are nullptrs, this is a degenerate poll reference. Return std::nullopt.
+    if (pindex_start == nullptr && pindex_end == nullptr) return std::nullopt;
+
+    // If the poll is still active, then pindex_end will be nullptr, so then use the current chain head.
+    if (pindex_end == nullptr) pindex_end = pindexBest;
+
+    // Since this calculation by its very nature is going to be heavyweight, we are going to
+    // dispense with using the heavyweight averaging functions, and instead accumulate the necessary
+    // values directly in a for loop that traverses the index. This will do it all in a single index
+    // pass, which is the most efficient.
+    double active_vote_weight_tally = 0;
+    double pool_magnitude = 0.0;
+    double network_magnitude = 0.0;
```

Comment:
FP math should not be used here in my opinion. While currently polls pass with AVW percentages higher than where FP errors should not be an issue, in the future polls just barely passing AVW requirements might cause controversies.

Also if in the future whitelisting is more automatized, it could affect the consensus.

Reply:
I tried int64_t interger math earlier and got overflows. I will do a range analysis in a minute...

Reply:
Definitely will need bignum math to do it. See...
![AVW_calcs-3](https://user-images.githubusercontent.com/7529186/124653051-d5cfd800-de6a-11eb-893a-45b74ff4d038.jpg)

Reply:
But, Difficulty itself is a double, and so is the constant in front. To try and do an all integer math equation, you have to go back to the bluepaper... I will do that next and see if I can do a successful integer only bignum calc.

Reply:
Ok. I have put up the bignum version. See the following to understand the calculation for netweight.
![AVW_calcs-4](https://user-images.githubusercontent.com/7529186/124674874-94025a00-de89-11eb-9cae-ed42e76c0728.jpg)

Note that I have not fully analyzed the overflow limits of the calculations, although the initial analysis indicates we are good for any reasonable poll and network data.

Reply:
I think the superblock problem for early superblocks is deeper than the flags. I think that height is around when we changed from a uncompressed to custom compressed superblock format. I will have to investigate this further.

Reply:
@div72 ... with the latest changes...
getpollresults df3af5f55268bd71b7498da4690d36b9ba7eb1f403162f72fce5d3cd950e04d9 
{
"poll_id": "df3af5f55268bd71b7498da4690d36b9ba7eb1f403162f72fce5d3cd950e04d9",
"poll_title": "technical_poll:_should_we_implement_fixed_pos_block_rewards?",
"poll_expired": true,
"starting_block_height": 1018333,
"ending_block_height": 1037461,
"votes": 170,
"invalid_votes": 19,
"total_weight": 58373158.79546228,
"active_vote_weight": 64808181.37995655,
"vote_percent_avw": 90.07066322881812,
"top_choice_id": 0,
"top_choice": "Yes",
"responses": [
{
"choice": "Yes",
"id": 0,
"weight": 54754202.42466300,
"votes": 122
},
{
"choice": "No",
"id": 1,
"weight": 2990099.75462796,
"votes": 16
},
{
"choice": "Abstain",
"id": 2,
"weight": 628856.61617132,
"votes": 13
}
]
}